test/python/pulse/test_transforms.py

```diff
@@ -401,12 +407,286 @@ def test_multiple_schedules(self):
+        # d1
+        reference = reference.insert(6, instructions.Delay(5, d1))
+        # d2
+        reference += instructions.Delay(11, d2)
+        self.assertEqual(schedule, reference)
+
+    def test_align_right_with_barrier(self):
+        """Test right alignment with a barrier."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+        d2 = pulse.DriveChannel(2)
+
+        schedule = pulse.Schedule()
+        schedule = schedule.insert(1, instructions.Delay(3, d0), mutate=True)
+        schedule.append(directives.RelativeBarrier(d0, d1, d2), mutate=True)
+        schedule = schedule.insert(17, instructions.Delay(11, d2), mutate=True)
+
+        sched_grouped = pulse.Schedule()
+        sched_grouped.insert(2, instructions.Delay(5, d1), mutate=True)
+        sched_grouped += instructions.Delay(7, d0)
+
+        schedule.append(sched_grouped, mutate=True)
+        schedule = transforms.remove_directives(transforms.align_right(schedule))
+
+        reference = pulse.Schedule()
+        # d0
+        reference = reference.insert(0, instructions.Delay(3, d0))
+        reference = reference.insert(7, instructions.Delay(7, d0))
+        # d1
+        reference = reference.insert(9, instructions.Delay(5, d1))
+        # d2
+        reference = reference.insert(3, instructions.Delay(11, d2))
+
+        self.assertEqual(schedule, reference)
+
+
+class TestGroup(QiskitTestCase):
+    """Test grouping transform."""
+
+    def test_group(self):
+        """Test grouping transform."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+
+        schedule = pulse.Schedule()
+        schedule += instructions.Delay(3, d0)
+
+        grouped = pulse.Schedule()
+        grouped += instructions.Delay(5, d1)
+        grouped += instructions.Delay(7, d0)
+        grouped = transforms.group(grouped)
+
+        schedule = transforms.align_left(schedule+grouped)
+
+        reference = pulse.Schedule()
+        # d0
+        reference += instructions.Delay(3, d0)
+        reference += instructions.Delay(7, d0)
+        # d1
+        reference = reference.insert(3, instructions.Delay(5, d1))
+        self.assertEqual(schedule, reference)
+
+
+class TestFlatten(QiskitTestCase):
+    """Test flattening transform."""
+
+    def test_flatten(self):
+        """Test the flatten transform."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+
+        schedule = pulse.Schedule()
+        schedule += instructions.Delay(3, d0)
+
+        grouped = pulse.Schedule()
+        grouped += instructions.Delay(5, d1)
+        grouped += instructions.Delay(7, d0)
+        # include a grouped schedule
+        grouped = transforms.group(grouped)
+        grouped = schedule+grouped
+
+        # flatten the schedule inline internal groups
+        flattened = transforms.flatten(grouped)
+
+        # align all the instructions to the left after flattening
+        schedule = transforms.align_left(flattened)
```

Comment:
How do you know that `align_left` cannot obfuscate the result of `flattened`? Is there a way to test this without using this other function on the schedule output?

Reply:
Verified aabe54bba.